### PR TITLE
Interactions: Add `step` function and support multiple levels of nesting

### PR DESCRIPTION
**Pull request by ghengeveld**
## What I did

Added a `step` function on the `play` function's `context` object. The `step` function effectively works like a Jest `describe` block, where you provide a label and a callback function as arguments:

```js
Demo.play = async ({ args, canvasElement, step }) => {
  await step('One', async () => {
    // do things
  })
  await step('Two', async () => {
    // do more things
  })
}
```

Screen Shot 2022-06-23 at 20 39 43

Note that for the time being it's not possible to debug interactions inside a `step` function (similar to `waitFor`). That's out of scope for this PR.

## How to test

- [x] Is this testable with Jest or Chromatic screenshots? yes
- [x] Does this need a new example in the kitchen sink apps? no
- [x] Does this need an update to the documentation? yes

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Comment by tmeasday** [comment]
Hmm, I'm not sure we want to add that dependency. I wonder if we should have a (basically hidden) mechanism where addons can add fields to the story/play context?

Although even if we do that there's a typing issue (how does `Story.play({ step })` get the right type?). It's tricky. WDYT @shilman?

**Comment by ghengeveld** [comment]
> Hmm, I'm not sure we want to add that dependency. I wonder if we should have a (basically hidden) mechanism where addons can add fields to the story/play context?

What would that mechanism look like?

**Comment by tmeasday** [comment]
OK, let's go for something in between:

 - We'll add an undocumented preset export `step` that ends up on the `projectAnnotations`, defaulting to `f => f()`
 - We'll add `step` to the `PlayFunctionContext` type
 - The preview will pass the `projectAnnotations.step` function to `play()`.

How does that sound @ghengeveld?

**Comment by ghengeveld** [comment]
> * We'll add an undocumented preset export `step` that ends up on the `projectAnnotations`, defaulting to `f => f()`

What happens if someone sets up another addon which exports a step value like that? Do we compose them?

**Comment by tmeasday** [comment]
I was thinking it's just last one wins with maybe a warning. (Thus the undocumented). 

Unless you think composing makes more sense?

**Comment by ghengeveld** [comment]
@wKich this might be of interest to you.

**Comment by ghengeveld** [comment]
This should be rebased off `future/base`

**Comment by yannbf** [comment]
One note about this change: it will break @storybook/testing-react and the like. The mechanism to get the step function ideally should be simple so it can be automated without the user doing anything

**Comment by tmeasday** [comment]
@yannbf if the default is `f => f()` I don't think it'll break anything if it isn't explicitly set anywhere?

**Comment by ghengeveld** [comment]
@tmeasday to make some changes to inject `step` into the `playContext` without having a dependency on `instrumenter`.

**Comment by tmeasday** [comment]
@ghengeveld my changes are merged now!

File: code/addons/interactions/src/Panel.tsx [hunk 1]
```diff
@@ -43,16 +43,19 @@ export const getInteractions = ({
...
+        childCallMap.set(ancestor, (childCallMap.get(ancestor) || []).concat(callId));
+      });
+      return visible;
     })
     .map(({ callId, status }) => ({ ...calls.get(callId), status } as Call))
     .map<Interaction>((call) => {
       const status =
         call.status === CallStates.ERROR &&
-        callsById.get(call.parentId)?.status === CallStates.ACTIVE
+        callsById.get(call.ancestors.slice(-1)[0])?.status === CallStates.ACTIVE
```
Comment (yannbf) [inline code/addons/interactions/src/Panel.tsx:58]: This looks a little dangerous. Can it be extracted to a local constant with some sort of safe check + default value?

File: code/addons/interactions/src/examples/Examples.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+export const WithSteps: Story = ({ onSubmit }) => (
+  <button type="button" onClick={() => onSubmit('clicked')}>
+    Click
+  </button>
+);
+WithSteps.play = async ({ args, canvasElement, step }) => {
+  step('Click button', async () => {
+    await userEvent.click(within(canvasElement).getByRole('button'));
+
+    step('Verify submit', async () => {
```
Comment (yannbf) [inline code/addons/interactions/src/examples/Examples.stories.tsx:113]: Shouldn't this be awaited? and the one below? This reminds me that once this is done we should add this check to the eslint plugin

File: code/lib/instrumenter/src/instrumenter.ts [hunk 3]
```diff
@@ -491,15 +493,15 @@ export class Instrumenter {
...
+          const { cursor, ancestors } = this.getState(call.storyId);
+          this.setState(call.storyId, { cursor: 0, ancestors: [...ancestors, call.id] });
+          const restore = () => this.setState(call.storyId, { cursor, ancestors });
 
           // Invoke the actual callback function.
           const res = arg(...args);
 
-          // Reset cursor and parentId to their original values before we entered the callback.
+          // Reset cursor and ancestors to their original values before we entered the callback.
           if (res instanceof Promise) res.then(restore, restore);
```
Comment (tmeasday) [inline code/lib/instrumenter/src/instrumenter.ts:505]: Should we be returning `res.then(...)` rather than just hooking it up and not doing anything with it?

File: code/lib/preview-web/src/render/StoryRender.ts [hunk 4]
```diff
@@ -24,6 +25,8 @@ import { Render, RenderType } from './Render';
 
 const { AbortController } = global;
 
+export type MaybePromise<T> = Promise<T> | T;
```
Comment (tmeasday) [inline code/lib/preview-web/src/render/StoryRender.ts:28]: Why did you move this in here I wonder?

File: code/lib/store/src/csf/composeConfigs.ts [hunk 5]
```diff
@@ -53,6 +54,6 @@ export function composeConfigs<TFramework extends AnyFramework>(
     render: getSingletonField(moduleExportList, 'render'),
     renderToDOM: getSingletonField(moduleExportList, 'renderToDOM'),
     applyDecorators: getSingletonField(moduleExportList, 'applyDecorators'),
-    runStep: composeStepRunners<TFramework>(getArrayField(moduleExportList, 'runStep')),
```
Comment (tmeasday) [inline code/lib/store/src/csf/composeConfigs.ts:56]: Clearly my tests here were bad, can you add one that failed with the old code?

File: code/lib/store/src/types.ts [hunk 6]
```diff
@@ -106,6 +106,14 @@ export type BoundStory<TFramework extends AnyFramework = AnyFramework> = Story<T
   storyFn: PartialStoryFn<TFramework>;
 };
 
+export declare type PlayContext<TFramework extends AnyFramework = AnyFramework> =
+  StoryContext<TFramework> & {
+    step: (
+      name: string,
+      play: (context: PlayContext<TFramework>) => MaybePromise<void>
+    ) => MaybePromise<void>;
+  };
```
Comment (tmeasday) [inline code/lib/store/src/types.ts:115]: This is defined in `@storybook/csf` already: https://github.com/ComponentDriven/csf/blob/f3e3044802f41c4885e8f564ed9ec140f6f845a6/src/story.ts#L116-L122

File: docs/snippets/common/storybook-interactions-play-function.js.mdx [hunk 7]
```diff
@@ -1,34 +1,39 @@
 ```js
 // MyForm.stories.js
 
-import { expect } from '@storybook/jest';
+import { expect } from "@storybook/jest"
```
[the remark this answers is not in the document]
Comment (ghengeveld) [reply docs/snippets/common/storybook-interactions-play-function.js.mdx:4]: Since the move to the `code` directory, there is no `.eslintrc` which applies `docs` anymore, so it used my VS Code settings instead. Makes sense.

[the remark this answers is not in the document]
Comment (ghengeveld) [reply docs/snippets/common/storybook-interactions-play-function.js.mdx:4]: And of course double quotes and no semi is the only proper way to format JS/TS. But alas.

File: code/addons/interactions/src/Panel.tsx (hunk 1 above)
Comment (ghengeveld) [reply code/addons/interactions/src/Panel.tsx:58]: `call.ancestors` is guaranteed to be an array (by TypeScript). Slicing an empty array always yields an empty array. Taking the first element of an empty array returns `undefined`, which we then check for using optional chaining. I don't see the problem unless you see a reason for `ancestors` to ever be null or undefined?

File: code/addons/interactions/src/examples/Examples.stories.tsx (hunk 2 above)
Comment (ghengeveld) [reply code/addons/interactions/src/examples/Examples.stories.tsx:113]: Yes absolutely 👍

File: docs/snippets/common/storybook-interactions-play-function.js.mdx (hunk 7 above)
[the remark this answers is not in the document]
Comment (ghengeveld) [reply docs/snippets/common/storybook-interactions-play-function.js.mdx:4]: Turns out it was `.prettierrc`. I added it and fixed the formatting.

File: code/lib/instrumenter/src/instrumenter.ts (hunk 3 above)
Comment (ghengeveld) [reply code/lib/instrumenter/src/instrumenter.ts:505]: We are returning `res` 2 lines down. It'll still have those callbacks attached.

File: code/lib/preview-web/src/render/StoryRender.ts (hunk 4 above)
Comment (ghengeveld) [reply code/lib/preview-web/src/render/StoryRender.ts:28]: Yeah it got migrated here when StoryRender was pulled out of PreviewWeb I think. Could just be in Preview, so I moved it there now.

File: code/lib/store/src/types.ts (hunk 6 above)
Comment (ghengeveld) [reply code/lib/store/src/types.ts:115]: Turns out we don't need it at all.

File: code/lib/instrumenter/src/instrumenter.ts (hunk 3 above)
Comment (tmeasday) [reply code/lib/instrumenter/src/instrumenter.ts:505]: It'll have the callbacks attached but a caller will not wait on the callbacks running. There's no guarantee the callbacks will run before the promise resolves. I guess it is not an issue but just a little detail that might trip something up down the track.

File: code/lib/store/src/csf/composeConfigs.ts (hunk 5 above)
Comment (ghengeveld) [reply code/lib/store/src/csf/composeConfigs.ts:56]: Done

Comment (tmeasday) [reply code/lib/store/src/csf/composeConfigs.ts:56]: Thank you!

**Review by tmeasday — APPROVED** [review]
LGTM!

**Comment by shilman** [comment]
@ghengeveld can you please review the chromatic change?

**Comment by ghengeveld** [comment]
The debugger seems broken right now. Need to investigate.

**Comment by ghengeveld** [comment]
@shilman This is good to go. Not sure what's up with cra-bench, seems like a Playwright or config issue unrelated to this PR.

**Comment by IanVS** [comment]
I suppose this needs to be added to the vite builder as well now, right?

**Comment by tmeasday** [comment]
Oh! Yeah definitely. Did we miss this too? 🤦
